File: nova_lang/semantic.py
"""
Semantic analysis for NovaLang: symbol tables, types, validation
"""

from .ast_nodes import *
from .token import TokenType

class SemanticError(Exception):
    pass

class Symbol:
    def __init__(self, name, typ):
        self.name = name
        self.type = typ

class FunctionSymbol:
    def __init__(self, name, params):
        self.name = name
        self.params = params  # list of param names
        # return type is num/text/flag, but here deduce later or store dynamic

class SemanticAnalyzer:
    def __init__(self):
        self.scopes = [{}]  # var scopes
        self.functions = {}
        self.in_loop = 0
        self.in_func = 0

    def error(self, msg):
        raise SemanticError(msg)

    def declare_var(self, name, typ):
        if name in self.scopes[-1]:
            self.error(f"Redeclaration of variable '{name}'")
        self.scopes[-1][name] = Symbol(name, typ)

    def lookup_var(self, name):
        for scope in reversed(self.scopes):
            if name in scope:
                return scope[name]
        self.error(f"Use of undeclared variable '{name}'")

    def analyze(self, node):
        return self.visit(node)

    def visit(self, node):
        method = 'visit_' + node.__class__.__name__
        if hasattr(self, method):
            return getattr(self, method)(node)
        else:
            # ignore nodes or basic types
            return None

    def visit_Program(self, node: Program):
        for stmt in node.statements:
            self.visit(stmt)

    def visit_VarDecl(self, node: VarDecl):
        vtyp = node.vartype  # str
        expr_type = self.visit(node.expr)
        if vtyp == 'num' and expr_type != 'num':
            self.error("Type mismatch: expected num")
        if vtyp == 'text' and expr_type != 'text':
            self.error("Type mismatch: expected text")
        if vtyp == 'flag' and expr_type != 'bool':
            self.error("Type mismatch: expected flag")
        self.declare_var(node.name, vtyp)

    def visit_Assign(self, node: Assign):
        sym = self.lookup_var(node.name)
        expr_type = self.visit(node.expr)
        if sym.type == 'num' and expr_type != 'num':
            self.error("Type mismatch in assignment to num")
        if sym.type == 'text' and expr_type != 'text':
            self.error("Type mismatch in assignment to text")
        if sym.type == 'flag' and expr_type != 'bool':
            self.error("Type mismatch in assignment to flag")

    def visit_Show(self, node: Show):
        self.visit(node.expr)

    def visit_Take(self, node: Take):
        sym = self.lookup_var(node.name)
        # allowed to read text/num/bool
        return sym.type

    def visit_When(self, node: When):
        # each condition must be bool
        for cond, stmts in node.cases:
            ctype = self.visit(cond)
            if ctype != 'bool':
                self.error("When condition must be boolean")
            self.scopes.append({})
            for s in stmts:
                self.visit(s)
            self.scopes.pop()
        if node.else_block:
            self.scopes.append({})
            for s in node.else_block:
                self.visit(s)
            self.scopes.pop()

    def visit_Loop(self, node: Loop):
        # loop i = expr to expr
        # i must be num
        start_t = self.visit(node.start_expr)
        end_t = self.visit(node.end_expr)
        if start_t != 'num' or end_t != 'num':
            self.error("Loop bounds must be num")
        # new scope
        self.scopes.append({})
        self.declare_var(node.var, 'num')
        self.in_loop += 1
        for s in node.body:
            self.visit(s)
        self.in_loop -= 1
        self.scopes.pop()

    def visit_Break(self, node: Break):
        if self.in_loop == 0:
            self.error("break outside loop")

    def visit_FuncDef(self, node: FuncDef):
        # Register function
        if node.name in self.functions:
            self.error(f"Redeclaration of function '{node.name}'")
        self.functions[node.name] = FunctionSymbol(node.name, node.params)

        # new scope
        self.scopes.append({})
        # declare params as num/text/flag? NovaLang spec not explicit, assume dynamic?
        # For simplicity, assume numeric parameters (or accept any?)
        # We'll accept num only? No: allow dynamic but check usage.
        # Here allow all types: treat params as 'num' by default? 
        # Given spec unclear, we treat them as untyped (None). We'll assign type upon first assignment?
        for p in node.params:
            self.declare_var(p, 'num')  # simplistic; can be improved

        self.in_func += 1
        for s in node.body:
            self.visit(s)
        # back expr
        ret_type = self.visit(node.back_expr)
        self.in_func -= 1
        self.scopes.pop()

    def visit_FuncCall(self, node: FuncCall):
        if node.name not in self.functions:
            self.error(f"Call to undeclared function '{node.name}'")
        func = self.functions[node.name]
        if len(node.args) != len(func.params):
            self.error(f"Function '{node.name}' expects {len(func.params)} arguments")
        # type check args
        for arg in node.args:
            self.visit(arg)
        # assume return type 'num' for simplicity
        return 'num'

    def visit_BinOp(self, node: BinOp):
        lt = self.visit(node.left)
        rt = self.visit(node.right)
        op = node.op.type
        # arithmetic
        if op in (TokenType.PLUS, TokenType.MINUS, TokenType.STAR, TokenType.SLASH):
            if lt == 'num' and rt == 'num':
                return 'num'
            if op == TokenType.PLUS and lt == 'text' and rt == 'text':
                return 'text'
            self.error("Invalid operands for arithmetic")
        # comparisons
        if op in (TokenType.GT, TokenType.LT, TokenType.GTEQ, TokenType.LTEQ, TokenType.EQEQ, TokenType.NOTEQ):
            # allow num-num or text-text for == or !=
            if lt == rt:
                return 'bool'
            self.error("Type mismatch in comparison")
        self.error("Unknown binary op")

    def visit_UnaryOp(self, node: UnaryOp):
        et = self.visit(node.expr)
        if node.op.type == TokenType.MINUS:
            if et == 'num':
                return 'num'
            self.error("Unary minus on non-num")
        return et

    def visit_Literal(self, node: Literal):
        return node.lit_type

    def visit_Identifier(self, node: Identifier):
        sym = self.lookup_var(node.name)
        return sym.type

File: nova_lang/semantic.cpp
#include "semantic.hpp"
#include <stdexcept>
#include <typeinfo>

SemanticAnalyzer::SemanticAnalyzer() {
    scopes.emplace_back();
}

void SemanticAnalyzer::enter_scope() { scopes.emplace_back(); }
void SemanticAnalyzer::exit_scope() { if (!scopes.empty()) scopes.pop_back(); }

void SemanticAnalyzer::declare_var(const std::string& name, const std::string& type) {
    if (scopes.back().count(name)) throw SemanticError("Redeclaration of variable '" + name + "'");
    scopes.back().emplace(name, Symbol(name, type));
}

Symbol* SemanticAnalyzer::lookup_var(const std::string& name) {
    for (auto it = scopes.rbegin(); it != scopes.rend(); ++it) {
        if (it->count(name)) return &((*it)[name]);
    }
    throw SemanticError("Use of undeclared variable '" + name + "'");
    return nullptr;
}

void SemanticAnalyzer::analyze(ASTNode* node) {
    visit(node);
}

std::string SemanticAnalyzer::visit(ASTNode* node) {
    if (!node) return "";
    if (auto p = dynamic_cast<Program*>(node)) return visit_Program(p);
    if (auto v = dynamic_cast<VarDecl*>(node)) return visit_VarDecl(v);
    if (auto a = dynamic_cast<Assign*>(node)) return visit_Assign(a);
    if (auto s = dynamic_cast<Show*>(node)) return visit_Show(s);
    if (auto t = dynamic_cast<Take*>(node)) return visit_Take(t);
    if (auto w = dynamic_cast<When*>(node)) return visit_When(w);
    if (auto lp = dynamic_cast<Loop*>(node)) return visit_Loop(lp);
    if (dynamic_cast<Break*>(node)) return visit_Break(dynamic_cast<Break*>(node));
    if (auto f = dynamic_cast<FuncDef*>(node)) return visit_FuncDef(f);
    if (auto fc = dynamic_cast<FuncCall*>(node)) return visit_FuncCall(fc);
    if (auto b = dynamic_cast<BinOp*>(node)) return visit_BinOp(b);
    if (auto u = dynamic_cast<UnaryOp*>(node)) return visit_UnaryOp(u);
    if (auto l = dynamic_cast<Literal*>(node)) return visit_Literal(l);
    if (auto id = dynamic_cast<Identifier*>(node)) return visit_Identifier(id);

    throw SemanticError("Unhandled AST node in semantic analyzer");
}

std::string SemanticAnalyzer::visit_Program(Program* node) {
    for (auto &s : node->statements) visit(s.get());
    return "";
}

std::string SemanticAnalyzer::visit_VarDecl(VarDecl* node) {
    std::string expr_t = visit(node->expr.get());
    std::string declared = node->vartype;
    if (declared == "num" && expr_t != "num") throw SemanticError("Type mismatch: expected num");
    if (declared == "text" && expr_t != "text") throw SemanticError("Type mismatch: expected text");
    if (declared == "flag" && expr_t != "bool") throw SemanticError("Type mismatch: expected flag");
    declare_var(node->name, declared);
    return declared;
}

std::string SemanticAnalyzer::visit_Assign(Assign* node) {
    Symbol* s = lookup_var(node->name);
    std::string expr_t = visit(node->expr.get());
    if (s->type == "num" && expr_t != "num") throw SemanticError("Type mismatch in assignment to num");
    if (s->type == "text" && expr_t != "text") throw SemanticError("Type mismatch in assignment to text");
    if (s->type == "flag" && expr_t != "bool") throw SemanticError("Type mismatch in assignment to flag");
    return s->type;
}

std::string SemanticAnalyzer::visit_Show(Show* node) {
    return visit(node->expr.get());
}

std::string SemanticAnalyzer::visit_Take(Take* node) {
    Symbol* s = lookup_var(node->name);
    return s->type;
}

std::string SemanticAnalyzer::visit_When(When* node) {
    for (auto &c : node->cases) {
        std::string ct = visit(c.first.get());
        if (ct != "bool") throw SemanticError("When condition must be boolean");
        enter_scope();
        for (auto &s : c.second) visit(s.get());
        exit_scope();
    }
    if (!node->else_block.empty()) {
        enter_scope();
        for (auto &s : node->else_block) visit(s.get());
        exit_scope();
    }
    return "";
}

std::string SemanticAnalyzer::visit_Loop(Loop* node) {
    std::string s1 = visit(node->start_expr.get());
    std::string s2 = visit(node->end_expr.get());
    if (s1 != "num" || s2 != "num") throw SemanticError("Loop bounds must be num");
    enter_scope();
    declare_var(node->var, "num");
    in_loop++;
    for (auto &st : node->body) visit(st.get());
    in_loop--;
    exit_scope();
    return "";
}

std::string SemanticAnalyzer::visit_Break(Break* /*node*/) {
    if (in_loop == 0) throw SemanticError("break outside loop");
    return "";
}

std::string SemanticAnalyzer::visit_FuncDef(FuncDef* node) {
    if (functions.count(node->name)) throw SemanticError("Redeclaration of function '" + node->name + "'");
    functions.emplace(node->name, FunctionSymbol(node->name, node->params));
    enter_scope();
    for (auto &p : node->params) declare_var(p, "num"); // simplistic; mark params as num
    in_func++;
    for (auto &s : node->body) visit(s.get());
    std::string ret = visit(node->back_expr.get());
    in_func--;
    exit_scope();
    return "";
}

std::string SemanticAnalyzer::visit_FuncCall(FuncCall* node) {
    if (!functions.count(node->name)) throw SemanticError("Call to undeclared function '" + node->name + "'");
    const auto &fs = functions.at(node->name);
    if (fs.params.size() != node->args.size()) throw SemanticError("Function '" + node->name + "' called with incorrect number of arguments");
    for (auto &a : node->args) visit(a.get());
    return "num";
}

std::string SemanticAnalyzer::visit_BinOp(BinOp* node) {
    std::string lt = visit(node->left.get());
    std::string rt = visit(node->right.get());
    auto op = node->op_type;
    if (op == TokenType::PLUS || op == TokenType::MINUS || op == TokenType::STAR || op == TokenType::SLASH) {
        if (lt == "num" && rt == "num") return "num";
        if (op == TokenType::PLUS && lt == "text" && rt == "text") return "text";
        throw SemanticError("Invalid operands for arithmetic");
    }
    if (op == TokenType::GT || op == TokenType::LT || op == TokenType::GTEQ || op == TokenType::LTEQ || op == TokenType::EQEQ || op == TokenType::NOTEQ) {
        if (lt == rt) return "bool";
        throw SemanticError("Type mismatch in comparison");
    }
    throw SemanticError("Unknown binary op");
}

std::string SemanticAnalyzer::visit_UnaryOp(UnaryOp* node) {
    std::string et = visit(node->expr.get());
    if (node->op_type == TokenType::MINUS) {
        if (et == "num") return "num";
        throw SemanticError("Unary minus on non-num");
    }
    return et;
}

std::string SemanticAnalyzer::visit_Literal(Literal* node) {
    return node->lit_type;
}

std::string SemanticAnalyzer::visit_Identifier(Identifier* node) {
    Symbol* s = lookup_var(node->name);
    return s->type;
}
